File: backend/app/models/onboarding_task.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base
from app.models.enums import OnboardingStage, TaskStatus


class OnboardingTask(Base):
    __tablename__ = "onboarding_tasks"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    employee_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("employees.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    bitrix_task_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    stage: Mapped[OnboardingStage] = mapped_column(nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    deadline_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    status: Mapped[TaskStatus] = mapped_column(default=TaskStatus.pending, nullable=False)
    is_required: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    vnd_document_code: Mapped[str | None] = mapped_column(String(64), nullable=True)
    external_link: Mapped[str | None] = mapped_column(String(1024), nullable=True)
    confirmation_required: Mapped[bool] = mapped_column(
        Boolean,
        default=False,
        nullable=False,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    completed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )

    employee = relationship("Employee", back_populates="tasks")

File: mobile/lib/models/onboarding_task.dart
/// Зеркало BackendDayOneTask из src/api/onboardingApi.ts.
class OnboardingTask {
  final int id;
  final int employeeId;
  final String stage;
  final String title;
  final String? description;
  final DateTime? deadlineAt;
  final String status; // pending | in_progress | completed | overdue
  final bool isRequired;
  final String? vndDocumentCode;
  final String? externalLink;
  final bool confirmationRequired;
  final String? documentUrl;
  final DateTime? completedAt;

  const OnboardingTask({
    required this.id,
    required this.employeeId,
    required this.stage,
    required this.title,
    this.description,
    this.deadlineAt,
    required this.status,
    required this.isRequired,
    this.vndDocumentCode,
    this.externalLink,
    this.confirmationRequired = false,
    this.documentUrl,
    this.completedAt,
  });

  bool get isCompleted => status == 'completed';

  factory OnboardingTask.fromJson(Map<String, dynamic> json) {
    return OnboardingTask(
      id: json['id'] as int,
      employeeId: json['employee_id'] as int,
      stage: json['stage'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      deadlineAt: json['deadline_at'] != null
          ? DateTime.tryParse(json['deadline_at'] as String)
          : null,
      status: (json['status'] as String?) ?? 'pending',
      isRequired: (json['is_required'] as bool?) ?? true,
      vndDocumentCode: json['vnd_document_code'] as String?,
      externalLink: json['external_link'] as String?,
      confirmationRequired: (json['confirmation_required'] as bool?) ?? false,
      documentUrl: json['document_url'] as String?,
      completedAt: json['completed_at'] != null
          ? DateTime.tryParse(json['completed_at'] as String)
          : null,
    );
  }
}
